lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2671,64 +2704,31 @@ SwiftLanguageRuntime::GetRuntimeUnwindPlan(ProcessSP process_sp,
-  // deref(reg-value + offset) yet, shortcut around it with
-  // a dwarf expression for now.
-  // The CFA of an async frame is the address of it's associated AsyncContext.
-  // In an async frame currently on the stack, this address is stored right
-  // before the saved frame pointer on the stack.
-  static const uint8_t g_cfa_dwarf_expression_x86_64[] = {
-      llvm::dwarf::DW_OP_breg6, // DW_OP_breg6, register 6 == rbp
-      0x78,                     //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-  };
-  static const uint8_t g_cfa_dwarf_expression_arm64[] = {
-      llvm::dwarf::DW_OP_breg29, // DW_OP_breg29, register 29 == fp
-      0x78,                      //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-  };
-
-  constexpr unsigned expr_size = sizeof(g_cfa_dwarf_expression_arm64);
-
-  static_assert(sizeof(g_cfa_dwarf_expression_x86_64) ==
-                    sizeof(g_cfa_dwarf_expression_arm64),
-                "Code relies on DWARF  expressions being the same size");
-
-  const uint8_t *expr = nullptr;
-  if (arch.GetMachine() == llvm::Triple::x86_64)
-    expr = g_cfa_dwarf_expression_x86_64;
-  else if (arch.GetMachine() == llvm::Triple::aarch64)
-    expr = g_cfa_dwarf_expression_arm64;
-  else
-    llvm_unreachable("Unsupported architecture");
-
   if (in_prologue) {
     if (indirect_context)
       row->GetCFAValue().SetIsRegisterDereferenced(regnums->async_ctx_regnum);
     else
       row->GetCFAValue().SetIsRegisterPlusOffset(regnums->async_ctx_regnum, 0);
   } else {
-    row->GetCFAValue().SetIsDWARFExpression(expr, expr_size);
+    // In indirect funclets, dereferencing (fp-8) once produces the CFA of the
+    // frame above. Dereferencing twice will produce the current frame's CFA.
+    bool with_double_deref = indirect_context;
+    llvm::ArrayRef<uint8_t> expr = GetAsyncRegFromFramePointerDWARFExpr(
+        arch.GetMachine(), with_double_deref);
+    row->GetCFAValue().SetIsDWARFExpression(expr.data(), expr.size());
   }
 
   if (indirect_context) {
     if (in_prologue) {
       row->SetRegisterLocationToSame(regnums->async_ctx_regnum, false);
     } else {
-      // In a "resume" coroutine, the passed context argument needs to be
-      // dereferenced once to get the context. This is reflected in the debug
-      // info so we need to account for it and report am async register value
-      // that needs to be dereferenced to get to the context.
-      // Note that the size passed for the DWARF expression is the size of the
-      // array minus one. This skips the last deref for this use.
-      assert(expr[expr_size - 1] == llvm::dwarf::DW_OP_deref &&
-             "Should skip a deref");
-      row->SetRegisterLocationToIsDWARFExpression(regnums->async_ctx_regnum,
-                                                  expr, expr_size - 1, false);
+      // In indirect contexts, the frame above needs to think their x22 is what
```

Comment:
This comment should not refer to all these ABI details?

Reply:
Mmmm I think your observation is a reflection of a bigger problem: there is a pile of nuanced tricks and non-standard unwinding behavior packed in this comment and code, all to make the dwarf expression evaluator be able to reason "entry value(x22) is the value of x22 in the parent frame". 

You might read this explanation  and point the obvious fact that the code comment doesn't even come close to saying that, and I agree :(  It took me weeks to realize what the original implementation was doing, it is difficult to capture _that_ into a small comment. I'll probably just remove this comment since my short-term goal is to delete most of this code and make this more a standard unwinding.